HSEARCH-5133 Support basic metrics aggregations (#4144)
hibernate/hibernate-search: documentation/src/main/asciidoc/public/reference/_search-dsl-aggregation.adoc

====

[NOTE]
====
You can always use the field type for the expected returned type.
Another option is to set `Double.class` for the expected returned type and you can always do it
even if the field is not double typed.
====

=== Min metric aggregation

The `min` aggregation provides the minimum value among the field values.

.Find the min release date among all the science fiction books
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/aggregation/AggregationDslIT.java[tags=min]
----
<1> Define the target field path to which you want to apply the aggregation function and the expected returned type.
====

=== Max metric aggregation

The `max` aggregation provides the maximum value among the field values.

.Find the max release date among all the science fiction books
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/aggregation/AggregationDslIT.java[tags=max]
----
<1> Define the target field path to which you want to apply the aggregation function and the expected returned type.
====

=== Count metric aggregation

> The `count` aggregation counts the field values.

Reviewer: It's
```suggestion
The `count` aggregation counts the number of documents in which the field is not empty. 
```

or not ? 🙈

> .Count anytime the price field has a value among all the science fiction books

Reviewer: ```suggestion
.Count the number of the science fiction books with prices
```

====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/aggregation/AggregationDslIT.java[tags=count]
----
<1> Define the target field path to which you want to apply the aggregation function. For this function a `Long.class` value is always returned.
====

=== Count distinct metric aggregation

> The `count distinct` aggregation counts the different field values.

Reviewer: ```suggestion
The `count distinct` aggregation counts the number of unique field values.
```
does it also count an empty value towards the total?

.Count anytime the price field has a different value among all the science fiction books
====
[source, JAVA, indent=0, subs="+callouts"]
----
include::{sourcedir}/org/hibernate/search/documentation/search/aggregation/AggregationDslIT.java[tags=count-distinct]
----
<1> Define the target field path to which you want to apply the aggregation function. For this function a `Long.class` value is always returned.
====

=== Avg metric aggregation

> Applies the SQL `avg` aggregation function on a given numeric or temporal field.

Reviewer: I think we've missed this one as we were replacing the sql-based definition in the previous iteration 🙈. Maybe
```suggestion
The `avg` aggregation calculates the average value of a given numeric or temporal field among the selected documents.
```

> .Compute the average of the prices all of the science fiction books

Reviewer: ```suggestion
.Compute the average price of all science fiction books
```
